### PR TITLE
GUVNOR-2502: Consolidate client-side bootstrap logic in various workbench implementations

### PR DESCRIPTION
**Pull request by paulovmr**

### CODE REVIEW

**Comment by paulovmr** [comment]
To merge together with:

https://github.com/droolsjbpm/droolsjbpm-build-bootstrap/pull/202
https://github.com/uberfire/uberfire/pull/372
https://github.com/droolsjbpm/kie-wb-common/pull/360
https://github.com/droolsjbpm/jbpm-form-modeler/pull/50
https://github.com/droolsjbpm/drools-wb/pull/176
https://github.com/droolsjbpm/jbpm-designer/pull/256
https://github.com/droolsjbpm/jbpm-console-ng/pull/421
https://github.com/droolsjbpm/optaplanner-wb/pull/69
https://github.com/droolsjbpm/livespark/pull/22
https://github.com/droolsjbpm/kie-wb-distributions/pull/285

**Comment by paulovmr** [comment]
jenkins please retest this

**Comment by manstis** [comment]
@wmedvede Could you please review this PR? I had a brief look and, from what I know about Docks, it appears OK.. but it'd be good to have someone who knows more about Docks review too.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+                                                  final boolean usersSystemActive ) {
+        final AbstractWorkbenchPerspectiveActivity defaultPerspective = getDefaultPerspectiveActivity();
+        final List<MenuItem> result = new ArrayList<>( 1 );
+
+        result.add( MenuFactory.newSimpleItem( constants.HomePage() )
+                            .place( new DefaultPlaceRequest( defaultPerspective.getIdentifier() ) )
+                            .endMenu()
+                            .build().getItems().get( 0 ) );
+
+        // Social menu items.
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java]: It's a matter of style, but I prefer methods extracted then comments.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+                iocManager.destroyBean( instance );
+            }
+        }
+
+        return defaultPerspective;
+    }
+
+    public List<PerspectiveActivity> getPerspectiveActivities() {
+
+        //Get Perspective Providers
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java]: same here.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        final Set<PerspectiveActivity> activities = activityManager.getActivities( PerspectiveActivity.class );
+
+        //Sort Perspective Providers so they're always in the same sequence!
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java]: same here.

File: kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/service/PlaceManagerActivityServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import javax.enterprise.context.ApplicationScoped;
+
+import org.jboss.errai.bus.server.annotations.Service;
+import org.kie.workbench.common.services.shared.service.PlaceManagerActivityService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Service
+@ApplicationScoped
+public class PlaceManagerActivityServiceImpl implements PlaceManagerActivityService {
```
Comment (ederign) [inline kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/service/PlaceManagerActivityServiceImpl.java]: I wonder if that should be moved to uberfire core? What do you think?

File: kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/service/PlaceManagerActivityServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    private List<String> availableActivities;
+
+    @PostConstruct
+    public void init() {
+    }
+
+    @Override
+    public void initActivities( List<String> activities ) {
+        this.availableActivities = activities;
```
Comment (ederign) [inline kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/service/PlaceManagerActivityServiceImpl.java]: what this method does?

**Comment by ederign** [comment]
Looks good for me. @paulovmr could you please take a look in my concerns?

**Comment by paulovmr** [comment]
@ederign Hi, I will take a look at this later (because I already changed my local context...). Thanks for the review!

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/docks/AuthoringWorkbenchDocks.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+            lastActiveContext = null;
+        }
+    }
+
+    private boolean shouldDisplayWestDocks( DataModelerContext context ) {
+        return context != null && context.getEditionMode() == DataModelerContext.EditionMode.GRAPHICAL_MODE;
+    }
+
+    private boolean hasPlannerDomainGrant() {
+        Set<String> grantedRoles = kieACL.getGrantedRoles( "wb_optaplanner_domain" );
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/docks/AuthoringWorkbenchDocks.java]: @manstis I wonder if there is a away to avoid this hard coded roles.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                                       KieWorkbenchACL kieACL,
+                                       ActivityBeansCache activityBeansCache ) {
+        this.appConfigService = appConfigService;
+        this.kieSecurityService = kieSecurityService;
+        this.pmas = pmas;
+        this.kieACL = kieACL;
+        this.activityBeansCache = activityBeansCache;
+    }
+
+    protected void customSecurityLoadedCallback() {
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java]: why instead of a empty method you don't set a real callback? (null as default)

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    @AfterInitialization
+    public void startDefaultWorkbench() {
+        kieSecurityService.call( new RemoteCallback<String>() {
+            public void callback( final String str ) {
+                initializeWorkbench( str );
+                customSecurityLoadedCallback();
+            }
+        } ).loadPolicy();
+
+        pmas.call().initActivities( activityBeansCache.getActivitiesById() );
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java]: @Override
 +    public void initActivities( List<String> activities ) {
 +        this.availableActivities = activities;
}

probably the method name is wrong.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            }
+        } ).loadPreferences();
+    }
+
+    void loadStyles() {
+        RoundedCornersResource.INSTANCE.roundCornersCss().ensureInjected();
+    }
+
+    void hideLoadingPopup() {
+        @SuppressWarnings("GwtToHtmlReferences")
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java]: @paulovmr in the future we should move this to uberfire.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+    public void addLogoutMenuItem() {
+        final Menus userMenus = MenuFactory.newTopLevelMenu( constants.LogOut() )
+                .respondsWith( new LogoutCommand() )
+                .endMenu()
+                .build();
+
+        userMenu.addMenus( userMenus );
+    }
+
+    public AbstractWorkbenchPerspectiveActivity getDefaultPerspectiveActivity() {
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java]: Maybe this also should be a Uberfire internal method, putting  "Home" menu as default perspective. (but probably in a new PR).

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+        return perspectives;
+    }
+
+    public List<Menus> getRoles() {
+        final Set<Role> roles = identity.getRoles();
+        final List<Menus> result = new ArrayList<>( roles.size() );
+
+        result.add( MenuFactory.newSimpleItem( constants.LogOut() ).respondsWith( new LogoutCommand() ).endMenu().build() );
+        for ( final Role role : roles ) {
+            if ( !role.getName().equals( "IS_REMEMBER_ME" ) ) {
```
Comment (ederign) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java]: @manstis @porcelli do you know why this?

**Comment by ederign** [comment]
@csadilek @paulovmr I've pointed out in the comments some duplicated code that we can extract and (maybe) move to UF core. We should talk about it later.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java (hunk 11 above)
Comment (manstis) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java]: @ederign @porcelli I believe it related to cookie based SSO (but I could be wrong - and I'm not sure whether it's even implemented any more, since we moved to Picket Link.. @porcelli any ideas?)

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/docks/AuthoringWorkbenchDocks.java (hunk 6 above)
Comment (manstis) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/docks/AuthoringWorkbenchDocks.java]: @ederign Hmmmmm.... some (probably should be *all*) Features/Groups are held [here](https://github.com/droolsjbpm/kie-wb-distributions/blob/master/kie-drools-wb/kie-drools-wb-home-page-community/src/main/java/org/kie/workbench/drools/client/security/KieWorkbenchFeatures.java) but then again we have (potentially) different ones for [product](https://github.com/droolsjbpm/kie-wb-distributions/blob/master/kie-drools-wb/kie-drools-wb-home-page-product/src/main/java/org/kie/workbench/drools/client/security/KieWorkbenchFeatures.java) and kie-wb vs kie-drools-wb.... so whilst we *could* improve on hard-coded ```Strings``` we still rely on hard-coded magic constants. Perhaps it could be reviewed... but TBH I'd wait for @dgutierr's replacement ACL to come along and see how this improves things (or by dropping his name here, he may even comment in advance!)

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java (hunk 1 above)
Comment (paulovmr) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java]: @ederign Methods extracted.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java (hunk 2 above)
Comment (paulovmr) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java]: @ederign I changed the variable name, I think it's enough in this case.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java (hunk 3 above)
Comment (paulovmr) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/menu/DefaultWorkbenchFeaturesMenusHelper.java]: @ederign I changed the variable name, I think it's enough in this case.

File: kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/service/PlaceManagerActivityServiceImpl.java (hunk 5 above)
Comment (paulovmr) [inline kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/service/PlaceManagerActivityServiceImpl.java]: @ederign From my understanding, this method caches client-side Uberfire Activities in the server-side. This service was extracted from jbpm-console-ng, and I think it could be promoted to Uberfire in another PR.

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java (hunk 7 above)
Comment (paulovmr) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java]: @ederign I've extended your suggestion to a list of callbacks, wdyt?

File: kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java (hunk 8 above)
Comment (paulovmr) [inline kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/entrypoint/DefaultWorkbenchEntryPoint.java]: @ederign As I mentioned in my other comment, I think this PlaceManagerActivityService could be promoted to Uberfire and improved a bit (in another PR), wdyt?

**Comment by csadilek** [comment]
+1
